src/main/java/org/jenkinsci/modules/windows_slave_installer/WindowsSlaveInstaller.java

```diff
@@ -133,22 +160,103 @@ public void run() {
+     * @param args slave.jar arguments to be passed
+     * @param extraMacroValues Additional macro values to be injected.
+     * @return Generated WinSW configuration file.
+     * @throws IOException The file cannot be generated
+     * @since TODO
+     */
+    @Restricted(NoExternalUse.class)
+    public static String generateSlaveXml(String id, String java, String vmargs, String args, @Nonnull Map<String, String> extraMacroValues) throws IOException {
+        // Just a legacy behavior for the obsolete installer
         String xml = IOUtils.toString(WindowsSlaveInstaller.class.getResourceAsStream("jenkins-slave.xml"), "UTF-8");
         xml = xml.replace("@ID@", id);
         xml = xml.replace("@JAVA@", java);
         xml = xml.replace("@VMARGS@", StringUtils.defaultString(vmargs));
         xml = xml.replace("@ARGS@", args);
         xml = xml.replace("\n","\r\n");
+        
+        for (Map.Entry<String, String> entry : extraMacroValues.entrySet()) {
+            xml = xml.replace("@" + entry.getKey() + "@", entry.getValue());
+        }
         return xml;
     }
+      
+    /*package*/ static String generateSlaveXml(String id, String java, @CheckForNull String vmargs, 
+                @Nonnull String args, @Nonnull Iterable<MacroValueProvider> providers
+            ) throws IOException {
+        Map<String, String> macroValues = new TreeMap<>();
+        for (MacroValueProvider provider : providers) {
+            //TODO: fail in the case of duplicated entries?
+            macroValues.putAll(provider.getMacroValues());
+        }
+        
+        return generateSlaveXml(id, java, vmargs, args, macroValues);
+    }
 
     private static final long serialVersionUID = 1L;
+    
+    /**
+     * Macro provider implementation for the internal use.
+     */
+    @Restricted(NoExternalUse.class)
+    /*package*/ static abstract class MacroValueProvider {
+       
+        @Nonnull
+        public abstract Map<String, String> getMacroValues();
+    }
+    
+    /*package*/ static class AgentURLMacroProvider extends MacroValueProvider {
+
+        @Nonnull
+        private final LaunchConfiguration launchConfiguration;
+        
+        public AgentURLMacroProvider(@Nonnull LaunchConfiguration launchConfig) {
+            this.launchConfiguration = launchConfig;
+        }
+
+        @Override
+        public Map<String, String> getMacroValues() {
+            Map<String, String> res = new TreeMap<>();
+            URL remotingURL = null;
+            try {
+                remotingURL = launchConfiguration.getLatestJarURL();
+            } catch (IOException ex) {
+                LOGGER.log(Level.SEVERE, "Failed to retrieve the latest Remoting JAR URL. Auto-download will be disabled", ex);
+            }
+            
+            final String macroValue;
+            if (remotingURL != null) {
+                macroValue = "<!-- <download from=\"" + remotingURL.toString() + "\" to=\"%BASE%\\slave.jar\"/> -->";
```

Comment:
I thought the agreement with @daniel-beck was that you only need to do this if the remotingURL scheme is not https.

(so long as a self signed cert does not blow up the update/service start on the agent...)